gcp/api/server.py

```diff
@@ -876,8 +876,24 @@ def query_by_version(context: QueryContext,
     bugs.extend(new_bugs)
 
   # Try querying by generic version for all cases.
-  new_bugs, _ = yield _query_by_generic_version(context, query, package_name,
-                                                ecosystem, purl, version)
+  new_bugs, generic_page_token = yield _query_by_generic_version(
+      context, query, package_name, ecosystem, purl, version)
+
+  if ecosystem:
+    if bugs and new_bugs:
+      # Both queries returned results, we can't support a page token.
+      next_page_token = None
+    elif new_bugs:
+      # The semver query returned no results, so we can return a page token
+      # here.
+      next_page_token = generic_page_token
```

Comment:
I'm not sure htis will work as is, because if we try this token against the semver query first, we may get an exception.

Let's just simplify all of this and drop support for pagination for SemVer for now. This logic is getting a little hard to follow. We can just move the `_query_by_generic_version`call into `_query_by_semver` and have it return no page tokens. We can restore the `_query_by_generic_version` call also in the `else` branch after line 866, and have that be the only case that returns a token.

Thinking more, there's potentially a way for us to support this better through our current semver_indexes, but either way it's going to be extremely unlikely to hit this pagination limit in practice.

Reply:
I am a bit concerned to move `_query_by_generic_version` call into `_query_by_semver`, so I:
 - restore the `_query_by_generic_version` call for non-semver ecosystem
 - instead of calling `_query_by_generic_version` once at the all for all cases, I add it to semver ecosystem to make the workflow clear

Basically I revoked https://github.com/google/osv.dev/pull/1783 but add an extra call of `_query_by_generic_version` for semver ecosystem.